lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class ClientError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config)
+      @config = config
+      @modified_options = {}
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+    end
+
+    attr_reader :config
+    attr_reader :scheduled_task, :promise # for specs
+
+    def start
+      return unless config.central_config?
+
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @promise =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @scheduled_task&.cancel
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 300..399
+        resp
+      when 400..499
+        raise ClientError, resp
+      when 500..599
+        raise ServerError, resp
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    def assign(update)
+      # For each updated option, store the original value,
+      # unless already stored
+      update.each_key do |key|
+        @modified_options[key] ||= config.get(key.to_sym)&.value
+      end
+
+      # If the new update doesn't set a previously modified option,
+      # revert it to the original
+      @modified_options.each_key do |key|
+        next if update.key?(key)
+        update[key] = @modified_options.delete(key)
+      end
```

Comment:
Maybe add one more test case when an option is set to nil just to make sure the existence of keys is always checked in the conditionals and not `hash[key]`?

Reply:
I’m not sure I follow?

Reply:
I was thinking about the scenario if someone wanted to "clear" out a setting they already modified. So using the example in the specs, setting ` { transaction_sample_rate: 0.5 }` then ` { transaction_sample_rate: nil }` 
I have only a surface-level understanding of this CentralConfig feature so perhaps that scenario is not relevant or worth testing.

Reply:
I think this would happen by removing the key in Kibana. So _un-setting_ means it's not sent anymore. On the contrary I could imagine someone in the future wanting to set an option to `nil` and how would we handle this if `nil` means _un-set_?

Reply:
If it's not possible now to set an option to `nil` then I guess testing for that is not necessary = )